demos/shadow/shadow_demo_main/shadow_demo_main.c

```diff
@@ -0,0 +1,644 @@
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+/**
+ * @file shadow_demo_main.c
+ * @brief Demo for showing how to use the Device Shadow library's API. This version
+ * of Device Shadow API provide macros and helper functions for assembling MQTT topics
+ * strings, and for determining whether an incoming MQTT message is related to the
+ * device shadow. The Device Shadow library does not depend on a MQTT library,
+ * therefore the code for MQTT connections are placed in another file (shadow_demo_helpers.c)
+ * to make it easy to read the code using Device Shadow library.
+ *
+ * This example assumes there is a powerOn state in the device shadow. It does the
+ * following operations:
+ * 1. Establish a MQTT connection by using the helper functions in shadow_demo_helpers.c.
+ * 2. Assemble strings for the MQTT topics of device shadow, by using macros defined by the Device Shadow library.
+ * 3. Subscribe to those MQTT topics by using helper functions in shadow_demo_helpers.c.
+ * 4. Publish a desired state of powerOn by using helper functions in shadow_demo_helpers.c.  That will cause
+ * a delta message to be sent to device.
+ * 5. Handle incoming MQTT messages in eventCallback, determine whether the message is related to the device
+ * shadow by using a function defined by the Device Shadow library (Shadow_MatchTopic). If the message is a
+ * device shadow delta message, set a flag for the main function to know, then the main function will publish
+ * a second message to update the reported state of powerOn.
+ * 6. Handle incoming message again in eventCallback. If the message is from update/accepted, verify that it
+ * has the same clientToken as previously published in the update message. That will mark the end of the demo.
+ */
+
+/* Standard includes. */
+#include <assert.h>
+#include <stdlib.h>
+#include <string.h>
+
+/* POSIX includes. */
+#include <unistd.h>
+
+#include "shadow_config.h"
+
+/* SHADOW API header. */
+#include "shadow.h"
+
+/* JSON API header. */
+#include "json.h"
+
+/* shadow demo helpers header. */
+#include "shadow_demo_helpers.h"
+
+
+/**
+ * @brief Format string representing a Shadow document with a "desired" state.
+ *
+ * The real json document will look like this:
+ * {
+ *   "state": {
+ *     "desired": {
+ *       "powerOn": 1
+ *     }
+ *   },
+ *   "clientToken": "021909"
+ * }
+ *
+ * Note the client token, which is required for all Shadow updates. The client
+ * token must be unique at any given time, but may be reused once the update is
+ * completed. For this demo, a timestamp is used for a client token.
```

Comment:
Do we have a reference for this?  I thought the client token was optional, and unique only with respect to in-flight requests from the same device.

Reply:
Yes, you are right, the client token is optional and we could take use of it to identify the each request from the same device.

Reply:
This [commit](https://github.com/aws/aws-iot-device-sdk-embedded-C/pull/1104/commits/49b40ee1665d75b8ed186c326788b6200655e313) should fix the comment.
please let me know if you have other questions.